File: app/ai/schemas/ai.py
"""
Wire format for /v1/ai/query.

Kept intentionally small: the v1 contract is one POST in, one JSON out.
Phase 2 will extend this with streaming and chat history; the response
shape here is the stable subset both phases will share, so the React
client can be built against it today.
"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

ChartType = Literal["line", "bar", "pie", "table"]

# Hard cap on conversational context sent back to us. Picked at 6 to
# keep prompt cost bounded (≈ 3 Q&A pairs) while giving the model
# enough recent context to resolve follow-ups like "and last week?".
# The router trims anything longer before it reaches the pipeline.
MAX_HISTORY_TURNS = 6


class ChatTurn(BaseModel):
    """One previous exchange in the conversation.

    ``content`` is the natural-language text only — for assistant turns
    we send the one-line summary, not the SQL or rows, to keep prompts
    small. The pipeline never re-uses prior SQL; it always regenerates
    from the catalog so guardrails apply uniformly to every call.
    """

    role: Literal["user", "assistant"]
    content: str = Field(..., min_length=1, max_length=4_000)


class AIQueryRequest(BaseModel):
    """User's natural-language question.

    ``tenant_id`` is OPTIONAL and only honored for superadmin principals.
    For all other roles the server uses the tenant from the JWT and
    rejects mismatched body values — see app/ai/routers/ai_query.py.

    ``history`` is the prior turns of the current chat session, oldest
    first. Optional — when empty the call behaves exactly like the v1
    one-shot endpoint. Server trims to the last ``MAX_HISTORY_TURNS``
    turns so a runaway client can't blow up the prompt budget.
    """

    question: str = Field(..., min_length=3, max_length=2_000)
    tenant_id: int | None = Field(
        default=None,
        description=(
            "Superadmin-only override. Tenant users have their tenant "
            "derived from the JWT; this field is rejected if it doesn't "
            "match the caller's tenant_id."
        ),
    )
    history: list[ChatTurn] = Field(
        default_factory=list,
        description=(
            "Prior turns of this chat session, oldest first. Capped to "
            f"the last {MAX_HISTORY_TURNS} turns server-side."
        ),
    )


class AIQueryError(BaseModel):
    """Structured error body. Returned with non-2xx responses where the
    failure happened *inside* the pipeline (e.g. SQL guard rejection)
    rather than during request parsing."""

    stage: Literal[
        "llm",
        "sql_guard",
        "tenant_filter",
        "execution",
        "summarization",
        "config",
    ]
    detail: str
    # Surface the offending SQL when the failure is the LLM's fault. This
    # is invaluable in production triage — without it, an "SQL rejected"
    # message is opaque. Safe to expose: tenant_id is server-bound so the
    # SQL never contains the literal value.
    sql: str | None = None


class AIQueryResponse(BaseModel):
    """Successful answer.

    ``rows`` is JSON-serialisable: every cell is one of int/float/str/bool/
    None/list/dict. Datetimes are stringified ISO-8601 by the executor.
    """

    question: str
    sql: str
    rows: list[dict[str, Any]]
    row_count: int
    columns: list[str]
    summary: str
    chart_type: ChartType
    # Lightweight debug payload — handy for the React "show the SQL"
    # toggle, opt-in surfacing in dev. Pure metadata; safe to return.
    provider: str
    model: str
    elapsed_ms: int
    # event_id of the audit row this answer was logged under. Surfaced so
    # the client can POST /ai/feedback referencing the exact query that
    # was rated, even after the user moves on to other questions.
    event_id: int | None = None


# ---------------------------------------------------------------------------
# History
# ---------------------------------------------------------------------------


# Cap the page size of GET /ai/history. Picked at 100 because the UI
# renders a sidebar list — anything bigger requires pagination anyway,
# and an unbounded scan would let a single user pin the audit table.
MAX_HISTORY_PAGE = 100


class AIHistoryItem(BaseModel):
    """One row of the user's own /v1/ai/query audit log.

    Includes any feedback the user has left on this query so the UI can
    render the right thumb state without a second round-trip.
    """

    event_id: int
    occurred_at: str  # ISO-8601 UTC
    question: str
    sql: str | None
    stage: str
    row_count: int | None
    elapsed_ms: int | None
    chart_type: ChartType | None
    provider: str | None
    model: str | None
    error_detail: str | None
    # -1 / 1 / None — None means "no feedback left yet"
    feedback_value: int | None = None
    feedback_comment: str | None = None


class AIHistoryResponse(BaseModel):
    items: list[AIHistoryItem]


# ---------------------------------------------------------------------------
# Feedback
# ---------------------------------------------------------------------------


class AIFeedbackRequest(BaseModel):
    """Thumbs up/down (+ optional comment) on a single past query.

    ``event_id`` references ``ai.query_log.event_id``. The server upserts
    on (user_id, event_id), so re-voting flips the value in place rather
    than appending — matches how the UI is wired (one thumb per message).
    """

    event_id: int = Field(..., ge=1)
    value: Literal[-1, 1]
    comment: str | None = Field(default=None, max_length=2_000)


class AIFeedbackResponse(BaseModel):
    feedback_id: int
    event_id: int
    value: int
    comment: str | None
    created_at: str


# ---------------------------------------------------------------------------
# Schema explorer
# ---------------------------------------------------------------------------


class AISchemaColumn(BaseModel):
    name: str
    type: str
    description: str = ""


class AISchemaTable(BaseModel):
    fqname: str
    schema_name: str  # split out so the UI can group by schema cheaply
    name: str
    description: str
    grain: str
    tenant_scoped: bool
    columns: list[AISchemaColumn]


class AISchemaResponse(BaseModel):
    """Snapshot of the curated catalog the SQL guard admits.

    This is a *whitelist*, not a live ``information_schema`` dump — see
    :mod:`app.ai.schemas.catalog` for why. The shape is stable across
    requests within a deploy; the client may cache it indefinitely.
    """

    tables: list[AISchemaTable]

File: web/src/api/ai.ts
// Typed client for the Text2SQL assistant.
//
// Backed by POST /v1/ai/query — see docs/ai_query.md for the full contract.
// Two design notes:
//
//   1. `chart_type` is one of four literals (line | bar | pie | table) chosen
//      server-side by a deterministic heuristic over the result shape, so the
//      client can switch on it without re-doing the inference.
//
//   2. The error response is structured: `{ stage, detail, sql? }`. The
//      `stage` discriminator tells the UI how to react (rephrase vs.
//      retry vs. page on-call) without parsing the human-readable message.
//      We surface it as `AIQueryError` so callers can `catch (e) { if
//      (e instanceof AIQueryError) ... }`.

import { ApiError, request } from "./client";

export type AIChartType = "line" | "bar" | "pie" | "table";

export type AIStage =
  | "llm"
  | "sql_guard"
  | "tenant_filter"
  | "execution"
  | "summarization"
  | "config";

/**
 * One prior turn of the current chat session. Sent back to the API so
 * the model can resolve follow-ups like "and last week?" against
 * earlier exchanges. Assistant `content` should be the summary text
 * only (not SQL/rows) — the backend doesn't reuse prior SQL.
 */
export interface ChatTurn {
  role: "user" | "assistant";
  content: string;
}

/** Server caps history to the last 6 turns; we trim client-side to match. */
export const MAX_HISTORY_TURNS = 6;

export interface AIQueryRequest {
  question: string;
  /** Superadmin-only override. Ignored / rejected for tenant users. */
  tenant_id?: number;
  /** Prior turns, oldest first. Optional — empty = one-shot Q&A. */
  history?: ReadonlyArray<ChatTurn>;
}

export interface AIQueryResponse {
  question: string;
  sql: string;
  /** JSON-coerced rows: every cell is one of string/number/boolean/null. */
  rows: ReadonlyArray<Record<string, unknown>>;
  row_count: number;
  columns: ReadonlyArray<string>;
  summary: string;
  chart_type: AIChartType;
  provider: string;
  model: string;
  elapsed_ms: number;
  /**
   * Audit log row id this answer was recorded under. Required to POST
   * feedback on this specific exchange. May be null if the server's
   * audit writer failed (fail-open) — in that case the UI hides the
   * feedback buttons for that message.
   */
  event_id: number | null;
}

/**
 * Structured error from POST /v1/ai/query. The server returns a body
 * shaped `{ detail: { stage, detail, sql? } }`; we re-throw it as this
 * class so React components can render the offending SQL when the
 * failure happened after generation.
 */
export class AIQueryError extends Error {
  stage: AIStage;
  sql: string | null;
  status: number;

  constructor(stage: AIStage, message: string, sql: string | null, status: number) {
    super(message);
    this.name = "AIQueryError";
    this.stage = stage;
    this.sql = sql;
    this.status = status;
  }
}

/** Re-shape the FastAPI error body into AIQueryError when we can. */
function toAIQueryError(err: unknown): never {
  if (err instanceof ApiError) {
    const body = err.body as { detail?: unknown } | null;
    const detail = body?.detail;
    if (
      detail &&
      typeof detail === "object" &&
      "stage" in detail &&
      typeof (detail as { stage: unknown }).stage === "string"
    ) {
      const d = detail as { stage: AIStage; detail?: string; sql?: string | null };
      throw new AIQueryError(d.stage, d.detail ?? err.message, d.sql ?? null, err.status);
    }
  }
  throw err;
}

export async function askAI(
  req: AIQueryRequest,
  signal?: AbortSignal,
): Promise<AIQueryResponse> {
  try {
    return await request<AIQueryResponse>("/ai/query", {
      method: "POST",
      body: req,
      signal,
    });
  } catch (e) {
    toAIQueryError(e);
  }
}

// ---------------------------------------------------------------------------
// History (GET /v1/ai/history)
// ---------------------------------------------------------------------------

/**
 * One row of the caller's own past /ai/query audit log, joined with any
 * feedback they've left. `event_id` is stable and FK-safe for POSTing
 * feedback back. `feedback_value` is null when no feedback was left.
 */
export interface AIHistoryItem {
  event_id: number;
  occurred_at: string; // ISO-8601 UTC
  question: string;
  sql: string | null;
  stage: string;
  row_count: number | null;
  elapsed_ms: number | null;
  chart_type: AIChartType | null;
  provider: string | null;
  model: string | null;
  error_detail: string | null;
  feedback_value: -1 | 1 | null;
  feedback_comment: string | null;
}

export interface AIHistoryResponse {
  items: ReadonlyArray<AIHistoryItem>;
}

/** Server caps at 100; the default fits the sidebar without paging. */
export async function fetchAIHistory(
  limit = 50,
  signal?: AbortSignal,
): Promise<AIHistoryResponse> {
  return request<AIHistoryResponse>(`/ai/history?limit=${limit}`, {
    method: "GET",
    signal,
  });
}

// ---------------------------------------------------------------------------
// Feedback (POST /v1/ai/feedback)
// ---------------------------------------------------------------------------

export type AIFeedbackValue = -1 | 1;

export interface AIFeedbackRequest {
  /** event_id from an AIQueryResponse — must belong to the caller. */
  event_id: number;
  value: AIFeedbackValue;
  comment?: string | null;
}

export interface AIFeedbackResponse {
  feedback_id: number;
  event_id: number;
  value: AIFeedbackValue;
  comment: string | null;
  created_at: string;
}

/**
 * Upsert the caller's feedback on a past /ai/query exchange. The server
 * upserts on (user_id, event_id), so calling this twice with different
 * values flips the vote in place rather than appending.
 *
 * Returns 404 if the event_id doesn't belong to the caller (stale local
 * state — the caller should drop its cached thumbs and refetch /history).
 */
export async function postAIFeedback(
  req: AIFeedbackRequest,
  signal?: AbortSignal,
): Promise<AIFeedbackResponse> {
  return request<AIFeedbackResponse>("/ai/feedback", {
    method: "POST",
    body: req,
    signal,
  });
}

// ---------------------------------------------------------------------------
// Schema explorer (GET /v1/ai/schema)
// ---------------------------------------------------------------------------

export interface AISchemaColumn {
  name: string;
  type: string;
  description: string;
}

export interface AISchemaTable {
  fqname: string;
  schema_name: string;
  name: string;
  description: string;
  grain: string;
  tenant_scoped: boolean;
  columns: ReadonlyArray<AISchemaColumn>;
}

export interface AISchemaResponse {
  tables: ReadonlyArray<AISchemaTable>;
}

/**
 * Snapshot of the curated catalog the SQL guard admits. Stable within a
 * deploy; cache aggressively (e.g. React Query `staleTime: Infinity`).
 */
export async function fetchAISchema(
  signal?: AbortSignal,
): Promise<AISchemaResponse> {
  return request<AISchemaResponse>("/ai/schema", {
    method: "GET",
    signal,
  });
}
